### PR TITLE
Sampling service

### PR DESCRIPTION
**Pull request by sbekele81**

### CODE REVIEW

File: xprof/xprof.rb.in [hunk 1]
```diff
@@ -767,19 +809,24 @@ def trace_and_on_node_processing(usr_argv)
     pids = if mpi_local_master?
              lm_setup_lttng(backends)
              lm_babeltrace(backends) if OPTIONS[:archive]
-            end
-
```
Comment (TApplencourt) [inline xprof/xprof.rb.in:771]: Can you run rubocopt (https://github.com/rubocop/rubocop)?

File: ze/Makefile.am [hunk 2]
```diff
@@ -172,23 +190,32 @@ libzetracepoints_la_CPPFLAGS = -I$(top_srcdir)/utils -I$(top_srcdir)/utils/inclu
...
-bt2dir = $(pkglibdir)/bt2
-bt2_LTLIBRARIES = libZEInterval.la
+lib_LTLIBRARIES = libze_loader.la libZEInterval.la
 
 nodist_libze_loader_la_SOURCES = \
 	$(ZE_PROBES_INCL) \
 	$(ZE_STATIC_PROBES_INCL) \
 	tracer_ze.c
-
+
```
Comment (TApplencourt) [inline ze/Makefile.am:199]: Please remove the empty line

File: ze/Makefile.am [hunk 3]
```diff
@@ -139,6 +140,8 @@ CLEANFILES += \
 	$(ZE_STATIC_PROBES_INCL) \
 	$(ZE_STATIC_PROBES_SRC)
 
+
+
 EXTRA_DIST += \
```
Comment (TApplencourt) [inline ze/Makefile.am:145]: Can you verify if the other `makefile.am` have some empty line?  We just want to be conscient

File: ze/Makefile.am [hunk 4]
```diff
@@ -172,23 +190,32 @@ libzetracepoints_la_CPPFLAGS = -I$(top_srcdir)/utils -I$(top_srcdir)/utils/inclu
...
+	$(LN_S) -f $(DESTDIR)$(libdir)/libze_loader.so.1.0.0 $(DESTDIR)$(pkglibdir)/ze/libze_loader.so.1
+	$(LN_S) -f $(DESTDIR)$(pkglibdir)/ze/libze_loader.so.1 $(DESTDIR)$(pkglibdir)/ze/libze_loader.so
+	$(MKDIR_P) $(DESTDIR)$(pkglibdir)/bt2
+	$(LN) -f $(DESTDIR)$(libdir)/libZEInterval.so $(DESTDIR)$(pkglibdir)/bt2/libZEInterval.so
+
+uninstall-hook:
+	$(RM) -f $(DESTDIR)$(pkglibdir)/ze/libze_loader.so.1 $(DESTDIR)$(pkglibdir)/ze/libze_loader.so
+	-rmdir $(DESTDIR)$(pkglibdir)/ze
+	$(RM) -f $(DESTDIR)$(pkglibdir)/bt2/libZEInterval.so
+	-rmdir $(DESTDIR)$(pkglibdir)/bt2
```
Comment (TApplencourt) [inline ze/Makefile.am:217]: We don't need that anymore

File: ze/tracer_ze_helpers.include.c [hunk 5]
```diff
@@ -1,4 +1,5 @@
-#include "thapi_sampling.h"
+//#include "thapi_sampling.h"
```
Comment (TApplencourt) [inline ze/tracer_ze_helpers.include.c:1]: Please just remove them :)

File: ze/sampling_daemon.h
Comment (TApplencourt) [inline ze/sampling_daemon.h:1]: Not sure we need this `.h` . Sampling_daemon should be a stand-aline

File: ze/tracer_ze_helpers.include.c [hunk 6]
```diff
@@ -789,484 +788,14 @@ static inline void _dump_memory_info(ze_command_list_handle_t hCommandList, cons
...
-}
 
 static void _load_tracer(void) {
   char *s = NULL;
   void *handle = NULL;
   int verbose = 0;
-  struct timespec interval;
-  thapi_sampling_init();
+  //struct timespec interval;
+  //thapi_sampling_init();
```
Comment (TApplencourt) [inline ze/tracer_ze_helpers.include.c:798]: Please remove

File: ze/sampling_daemon.c [hunk 7]
```diff
@@ -0,0 +1,803 @@
...
+#include <unistd.h>
+#include <signal.h>
+#include <unistd.h>
+#include <stdbool.h>
+
+#define SIG_SAMPLING_READY (SIGRTMIN)
+#define SIG_SAMPLING_FINISH (SIGRTMIN + 1)
+
+#define ZES_INIT_PTR zesInit_ptr
+
```
Comment (TApplencourt) [inline ze/sampling_daemon.c:21]: Please remove the empty line :)

File: ze/sampling_daemon.c [hunk 8]
```diff
@@ -0,0 +1,803 @@
...
+  if (!handle) {
+    _DL_ERROR_MSG();
+    return 1;
+  }
+  //Find zes symbols
+  find_ze_symbols(handle, verbose);
+  //Initialize device and telemetry handles
+  initializeHandles();
+  // Run the signal loop
+  signal(SIG_SAMPLING_FINISH, signal_handler);
```
Comment (TApplencourt) [inline ze/sampling_daemon.c:792]: I think you send `SIG_SAMPLING_READY` not finish

File: ze/sampling_daemon.c [hunk 9]
```diff
@@ -0,0 +1,803 @@
...
+  }
+
+  if (!handle) {
+    _DL_ERROR_MSG();
+    return 1;
+  }
+  //Find zes symbols
+  find_ze_symbols(handle, verbose);
+  //Initialize device and telemetry handles
+  initializeHandles();
```
Comment (TApplencourt) [inline ze/sampling_daemon.c:790]: You need to send `Ready` have `initializeHandles` no?

File: ze/sampling_daemon.c [hunk 10]
```diff
@@ -0,0 +1,803 @@
+#include "sampling_daemon.h"
+#include "../sampling/thapi_sampling.h"
```
Comment (TApplencourt) [inline ze/sampling_daemon.c:2]: I think it should just be `thapi_sampling.h` and the autotool should just add `sampling`  to `-I` IMO.

What do you think @Kerilk ?

File: xprof/xprof.rb.in [hunk 11]
```diff
@@ -732,9 +732,49 @@ def gm_rename_folder
...
+    @pid = spawn("#{daemon_path} #{parent_pid}")
+    Process.detach(@pid)
+
+    wait_for_ready_signal
+  end
+
+  def finalize
+    return unless @pid
+
+    Process.kill(SIG_SAMPLING_FINISH, @pid)
```
Comment (TApplencourt) [inline xprof/xprof.rb.in:760]: Finalize should wait for Read too

File: ze/sampling_daemon.c [hunk 12]
```diff
@@ -0,0 +1,803 @@
...
+    thapi_unregister_sampling(_sampling_handle);
+    _sampling_handle = NULL;
+  }
+}
+
+void signal_handler(int signum) {
+  if (signum == SIG_SAMPLING_FINISH) {
+    cleanup_sampling();
+    running = false;
+  }
```
[the remark this answers is not in the document]
Comment (sbekele81) [reply ze/sampling_daemon.c:757]: What I did is that once invoked the samling_daemon do initialization and sends READY to the parent. Then it goes into the loop and waits until finish.

File: xprof/xprof.rb.in (hunk 11 above)
Comment (sbekele81) [reply xprof/xprof.rb.in:760]: I included the wait there too. Thanks.

File: ze/sampling_daemon.h
Comment (sbekele81) [reply ze/sampling_daemon.h:1]: removed

File: ze/Makefile.am (hunk 2 above)
Comment (sbekele81) [reply ze/Makefile.am]: Removed

File: ze/Makefile.am (hunk 3 above)
Comment (sbekele81) [reply ze/Makefile.am]: Removed

File: ze/Makefile.am (hunk 4 above)
Comment (sbekele81) [reply ze/Makefile.am]: Done

File: ze/sampling_daemon.c (hunk 9 above)
Comment (sbekele81) [reply ze/sampling_daemon.c:772]: I send READY after initializing 
if (kill(parent_pid, SIG_SAMPLING_READY) != 0) {
    _ERROR_MSG("Failed to send READY signal to parent");
  }

File: ze/sampling_daemon.c [hunk 13]
```diff
@@ -0,0 +1,803 @@
...
+        readFabricPorts_dump(driverIdx, deviceIdx);
+      }
+      if (tracepoint_enabled(lttng_ust_ze_sampling, memStats)) {
+        readMemModules_dump(driverIdx, deviceIdx);
+      }
+    }
+  }
+}
+void process_sampling() {
+
```
[the remark this answers is not in the document]
Comment (sbekele81) [reply ze/sampling_daemon.c:717]: Removed, thanks.

**Comment by TApplencourt** [comment]
The CI failed with

```
# /home/runner/work/THAPI/THAPI/build/ici/bin/babeltrace_thapi:195:in `block in get_and_add_components': ze plugins was not found (RuntimeError)
# 	from /home/runner/work/THAPI/THAPI/build/ici/bin/babeltrace_thapi:192:in `each'
# 	from /home/runner/work/THAPI/THAPI/build/ici/bin/babeltrace_thapi:192:in `filter_map'
# 	from /home/runner/work/THAPI/THAPI/build/ici/bin/babeltrace_thapi:192:in `get_and_add_components'
# 	from /home/runner/work/THAPI/THAPI/build/ici/bin/babeltrace_thapi:314:in `no_gc'
# 	from /home/runner/work/THAPI/THAPI/build/ici/bin/babeltrace_thapi:329:in `create_and_run_graph'
# 	from /home/runner/work/THAPI/THAPI/build/ici/bin/babeltrace_thapi:461:in `<main>'
# I, [2024-11-15T15:02:33.254666 #6219]  INFO -- : IO dump: exit
```

File: xprof/xprof.rb.in [hunk 14]
```diff
@@ -754,9 +796,11 @@ def trace_and_on_node_processing(usr_argv)
     end
     # we can kill the session daemon
     lm_lttng_kill_sessiond
+    sampling_daemon&.finalize
```
Comment (TApplencourt) [inline xprof/xprof.rb.in:799]: what does this `&` doing? Never saw it before

Comment (sbekele81) [reply xprof/xprof.rb.in:799]: It is safe navigation operator.  Used it to avoid error in case the object is nil.

File: ze/Makefile.am [hunk 15]
```diff
@@ -172,11 +185,7 @@ libzetracepoints_la_CPPFLAGS = -I$(top_srcdir)/utils -I$(top_srcdir)/utils/inclu
 libzetracepoints_la_CFLAGS = -fPIC -Wall -Wextra -Wno-unused-parameter -Wno-type-limits -Wno-sign-compare $(WERROR) $(LTTNG_UST_CFLAGS)
 libzetracepoints_la_LDFLAGS = $(LTTNG_UST_LIBS)
 
-zedir = $(pkglibdir)/ze
-ze_LTLIBRARIES = libze_loader.la
-
-bt2dir = $(pkglibdir)/bt2
-bt2_LTLIBRARIES = libZEInterval.la
```
Comment (TApplencourt) [inline ze/Makefile.am:179]: The `ze` failing may be caused by said

Comment (sbekele81) [reply ze/Makefile.am:179]: Will do that. Thanks.

**Comment by TApplencourt** [comment]
Ok perfect! Merging! gj

**Review by TApplencourt — APPROVED** [review]

**Pull request merged**